com.unity.multiplayer.mlapi/Runtime/Metrics/NetworkMetrics.cs

```diff
@@ -0,0 +1,154 @@
+        {
+            m_NetworkManager = networkManager;
+            Dispatcher = new MetricDispatcherBuilder()
+                .WithMetricEvents(m_NamedMessageSentEvent, m_NamedMessageReceivedEvent)
+                .WithMetricEvents(m_UnnamedMessageSentEvent, m_UnnamedMessageReceivedEvent)
+                .WithMetricEvents(m_NetworkVariableDeltaSentEvent, m_NetworkVariableDeltaReceivedEvent)
+                .WithMetricEvents(m_ObjectSpawnSentEvent, m_ObjectSpawnReceivedEvent)
+                .WithMetricEvents(m_ObjectDestroySentEvent, m_ObjectDestroyReceivedEvent)
+                .Build();
+
+            Dispatcher.RegisterObserver(MLAPIObserver.Observer);
+        }
+
+        internal IMetricDispatcher Dispatcher { get; }
+
+        public void TrackNetworkObject(NetworkObject networkObject)
+        {
+            if (!m_NetworkGameObjects.ContainsKey(networkObject.NetworkObjectId))
+            {
+                m_NetworkGameObjects[networkObject.NetworkObjectId] = new NetworkObjectIdentifier(networkObject.name, networkObject.NetworkObjectId);
+            }
+        }
+
+        public void TrackNamedMessageSent(ulong receiverClientId, string messageName, ulong bytesCount)
+        {
+            m_NamedMessageSentEvent.Mark(new NamedMessageEvent(new ConnectionInfo(receiverClientId), messageName, bytesCount));
+        }
+
+        public void TrackNamedMessageSent(IReadOnlyCollection<ulong> receiverClientIds, string messageName, ulong bytesCount)
+        {
+            foreach (var receiver in receiverClientIds)
+            {
+                TrackNamedMessageSent(receiver, messageName, bytesCount);
+            }
+        }
+
+        public void TrackNamedMessageReceived(ulong senderClientId, string messageName, ulong bytesCount)
+        {
+            m_NamedMessageReceivedEvent.Mark(new NamedMessageEvent(new ConnectionInfo(senderClientId), messageName, bytesCount));
+        }
+
+        public void TrackUnnamedMessageSent(ulong receiverClientId, ulong bytesCount)
+        {
+            m_UnnamedMessageSentEvent.Mark(new UnnamedMessageEvent(new ConnectionInfo(receiverClientId), bytesCount));
+        }
+
+        public void TrackUnnamedMessageSent(IReadOnlyCollection<ulong> receiverClientIds, ulong bytesCount)
+        {
+            foreach (var receiverClientId in receiverClientIds)
+            {
+                TrackUnnamedMessageSent(receiverClientId, bytesCount);
+            }
+        }
+
+        public void TrackUnnamedMessageReceived(ulong senderClientId, ulong bytesCount)
+        {
+            m_UnnamedMessageReceivedEvent.Mark(new UnnamedMessageEvent(new ConnectionInfo(senderClientId), bytesCount));
+        }
+
+        public void TrackNetworkVariableDeltaSent(ulong receiverClientId, ulong networkObjectId, string gameObjectName, string variableName, ulong bytesCount)
+        {
+            variableName = PrettyPrintVariableName(variableName);
```

Comment:
We should take another look at this, it is doing a runtime allocation that will have the same result for each network variable